### DISCUSSION TITLE
Filter/middleware not working well with streaming

### DISCUSSION THREAD

**Discussion by kyleqian** [sinatra/sinatra#1912] [Q&A]
I noticed that streaming doesn't work well with filters or middleware. In the following snippet, both `stream block 1` and `stream block 2` get printed twice. I think this somewhat defeats the purpose of streaming - for instance, if we want to stream something first which doesn't require computation, then stream something requiring computation next. Is there a way for everything in the stream block to only be run once overall? I'd expect middleware to only run once per write.

```ruby
  after do
    response.body.each do |str|
      # Do things
    end
  end

  get '/stream' do
    stream do |out|
      puts "stream block 1"
      out << "stream block 1"
      puts "stream block 2"
      out << "stream block 2"
    end
  end
```

```
stream block 1
stream block 2
stream block 1
stream block 2
```

**Comment by dentarg** [comment]
> I'd expect middleware to only run once per write

That's not what's documented at https://github.com/sinatra/sinatra/tree/fb76534810a1b3d6365eb8e2296bf627cfbef8de#filters

> After filters are evaluated after each request within the same context as the routes will be and can also modify the request and response. Instance variables set in before filters and routes are accessible by after filters:

I don't think it makes sense to combine use of `after` and `stream`, don't you want the client to start consume your `stream` as soon as possible?
